---
title: "SFS-based analyses"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, eval=F, message=F)
```

```{r eval=T}
library(tidyverse)
library(ggstatsplot)
library(cowplot)
```

## Heterozygosity

#### With transitions

```{bash eval=F}
## Relaxed quality filter of 20
nohup nice -n 19 bash /workdir/genomic-data-analysis/scripts/get_heterozygosity.sh \
/workdir/batch-effect/ \
/workdir/batch-effect/sample_lists/bam_list_realigned.txt \
/workdir/cod/reference_seqs/gadMor3.fasta \
2 \
10 \
20 \
30 \
> /workdir/batch-effect/nohups/get_heterozygosity.nohup &
## More stringent quality filter of 33
nohup nice -n 19 bash /workdir/genomic-data-analysis/scripts/get_heterozygosity.sh \
/workdir/batch-effect/ \
/workdir/batch-effect/sample_lists/bam_list_realigned.txt \
/workdir/cod/reference_seqs/gadMor3.fasta \
2 \
10 \
33 \
30 \
> /workdir/batch-effect/nohups/get_heterozygosity_stringent.nohup &
```

#### Without transitions

```{bash eval=F}
## Relaxed quality filter of 20
nohup bash /workdir/cod/greenland-cod/scripts/get_heterozygosity_notrans.sh \
/workdir/batch-effect/ \
/workdir/batch-effect/sample_lists/bam_list_realigned.txt \
/workdir/cod/reference_seqs/gadMor3.fasta \
2 \
10 \
20 \
30 \
> /workdir/batch-effect/nohups/get_heterozygosity_notrans.nohup &
## More stringent quality filter of 33
nohup nice -n 19 bash /workdir/cod/greenland-cod/scripts/get_heterozygosity_notrans.sh \
/workdir/batch-effect/ \
/workdir/batch-effect/sample_lists/bam_list_realigned.txt \
/workdir/cod/reference_seqs/gadMor3.fasta \
2 \
10 \
33 \
30 \
> /workdir/batch-effect/nohups/get_heterozygosity_notrans_stringent.nohup &
```

#### Rerun a few outlier individuals 

The optimization algorithm may have stuck at a local optimum for these individuals so I'll give them another try

```{bash eval=FALSE}
## Relaxed quality filter of 20, including transitions
echo "/workdir/batch-effect/bam/PAA2011_728_55151_7_pe_bt2_gadMor3_sorted_dedup_overlapclipped_realigned.bam
/workdir/batch-effect/bam/UUM2010_048_55174_7_pe_bt2_gadMor3_sorted_dedup_overlapclipped_realigned.bam
/workdir/batch-effect/bam/BUK2011_057_55110_7_pe_bt2_gadMor3_sorted_dedup_overlapclipped_realigned.bam
/workdir/batch-effect/bam/KNG2011_397_14247X129_4_se_bt2_gadMor3_sorted_dedup_realigned.bam" \
> /workdir/batch-effect/sample_lists/bam_list_realigned_rerun_heterozygosity_1.txt
nohup nice -n 19 bash /workdir/genomic-data-analysis/scripts/get_heterozygosity.sh \
/workdir/batch-effect/ \
/workdir/batch-effect/sample_lists/bam_list_realigned_rerun_heterozygosity_1.txt \
/workdir/cod/reference_seqs/gadMor3.fasta \
2 \
10 \
20 \
30 \
> /workdir/batch-effect/nohups/rerun_heterozygosity_1.nohup &
## Stringent quality filter of 33, including transitions
echo "/workdir/batch-effect/bam/BUK2011_015_55230_7_pe_bt2_gadMor3_sorted_dedup_overlapclipped_realigned.bam
/workdir/batch-effect/bam/ITV2011_782_55236_7_pe_bt2_gadMor3_sorted_dedup_overlapclipped_realigned.bam
/workdir/batch-effect/bam/BUK2011_051_55157_7_pe_bt2_gadMor3_sorted_dedup_overlapclipped_realigned.bam
/workdir/batch-effect/bam/KNG2011_397_14247X129_4_se_bt2_gadMor3_sorted_dedup_realigned.bam" \
> /workdir/batch-effect/sample_lists/bam_list_realigned_rerun_heterozygosity_2.txt
nohup nice -n 19 bash /workdir/genomic-data-analysis/scripts/get_heterozygosity.sh \
/workdir/batch-effect/ \
/workdir/batch-effect/sample_lists/bam_list_realigned_rerun_heterozygosity_2.txt \
/workdir/cod/reference_seqs/gadMor3.fasta \
2 \
10 \
33 \
30 \
> /workdir/batch-effect/nohups/rerun_heterozygosity_2.nohup &
## Relaxed quality filter of 20, excluding transitions
echo "/workdir/batch-effect/bam/ITV2011_672_55181_7_pe_bt2_gadMor3_sorted_dedup_overlapclipped_realigned.bam
/workdir/batch-effect/bam/PAA2011_708_55139_7_pe_bt2_gadMor3_sorted_dedup_overlapclipped_realigned.bam
/workdir/batch-effect/bam/BUK2011_029_55108_7_pe_bt2_gadMor3_sorted_dedup_overlapclipped_realigned.bam
/workdir/batch-effect/bam/KNG2011_413_55159_7_pe_bt2_gadMor3_sorted_dedup_overlapclipped_realigned.bam
/workdir/batch-effect/bam/UUM2010_038_55119_7_pe_bt2_gadMor3_sorted_dedup_overlapclipped_realigned.bam
/workdir/batch-effect/bam/KNG2011_377_55155_7_pe_bt2_gadMor3_sorted_dedup_overlapclipped_realigned.bam" \
> /workdir/batch-effect/sample_lists/bam_list_realigned_rerun_heterozygosity_3.txt
nohup bash /workdir/cod/greenland-cod/scripts/get_heterozygosity_notrans.sh \
/workdir/batch-effect/ \
/workdir/batch-effect/sample_lists/bam_list_realigned_rerun_heterozygosity_3.txt \
/workdir/cod/reference_seqs/gadMor3.fasta \
2 \
10 \
20 \
30 \
> /workdir/batch-effect/nohups/rerun_heterozygosity_3.nohup &
## Stringent quality filter of 33, excluding transitions
echo "/workdir/batch-effect/bam/NAR2008_002_55156_7_pe_bt2_gadMor3_sorted_dedup_overlapclipped_realigned.bam
/workdir/batch-effect/bam/ATP2011_118_14247X184_6_se_bt2_gadMor3_sorted_dedup_realigned.bam
/workdir/batch-effect/bam/IKE2011_978_55137_7_pe_bt2_gadMor3_sorted_dedup_overlapclipped_realigned.bam
/workdir/batch-effect/bam/QQL2011_860_55166_7_pe_bt2_gadMor3_sorted_dedup_overlapclipped_realigned.bam
/workdir/batch-effect/bam/BUK2011_045_55130_7_pe_bt2_gadMor3_sorted_dedup_overlapclipped_realigned.bam
/workdir/batch-effect/bam/ITV2011_714_55194_7_pe_bt2_gadMor3_sorted_dedup_overlapclipped_realigned.bam
/workdir/batch-effect/bam/UUM2010_038_55119_7_pe_bt2_gadMor3_sorted_dedup_overlapclipped_realigned.bam
/workdir/batch-effect/bam/QQL2011_886_55196_7_pe_bt2_gadMor3_sorted_dedup_overlapclipped_realigned.bam"\
> /workdir/batch-effect/sample_lists/bam_list_realigned_rerun_heterozygosity_4.txt
nohup nice -n 19 bash /workdir/cod/greenland-cod/scripts/get_heterozygosity_notrans.sh \
/workdir/batch-effect/ \
/workdir/batch-effect/sample_lists/bam_list_realigned_rerun_heterozygosity_4.txt \
/workdir/cod/reference_seqs/gadMor3.fasta \
2 \
10 \
33 \
30 \
> /workdir/batch-effect/nohups/rerun_heterozygosity_4.nohup &
## Rerun five samples prior to sliding window trimming (i.e. in the /workdir/cod/greenland-cod directory)
echo "/workdir/cod/greenland-cod/angsd/heterozygosity/UUM2010_036_55109_7_pe_bt2_gadMor3_minq20_sorted_dedup_overlapclipped_realigned_mindp2_maxdp10_minq20_minmapq30
/workdir/cod/greenland-cod/angsd/heterozygosity/UUM2010_096_55133_7_pe_bt2_gadMor3_minq20_sorted_dedup_overlapclipped_realigned_mindp2_maxdp10_minq20_minmapq30
/workdir/cod/greenland-cod/angsd/heterozygosity/ATP2011_118_14247X184_6_se_bt2_gadMor3_minq20_sorted_dedup_realigned_mindp2_maxdp10_minq20_minmapq30
/workdir/cod/greenland-cod/angsd/heterozygosity/QQL2011_888_55216_7_pe_bt2_gadMor3_minq20_sorted_dedup_overlapclipped_realigned_mindp2_maxdp10_minq20_minmapq30
/workdir/cod/greenland-cod/angsd/heterozygosity/ATP2011_111_55173_7_pe_bt2_gadMor3_minq20_sorted_dedup_overlapclipped_realigned_mindp2_maxdp10_minq20_minmapq30" \
> /workdir/batch-effect/sample_lists/saf_list_rerun_heterozygosity.txt
for LINE in `cat /workdir/batch-effect/sample_lists/saf_list_rerun_heterozygosity.txt`; do
  /workdir/programs/angsd0.931/angsd/misc/realSFS \
  $LINE'.saf.idx' \
  -P 4 \
  > $LINE'.ml' &
done
```

#### Summarize and plot

###### Genome-wide, relaxed vs. stringent mapping quality filter and including vs. excluding transitions

```{r eval=TRUE, fig.width=10, fig.height=8}
sample_table <- read_tsv("../sample_lists/sample_table_merged.tsv")
for (i in 1:nrow(sample_table)){
  sample_seq_id <- sample_table$sample_seq_id[i]
  sample_id <- sample_table$sample_id_corrected[i]
  population <- sample_table$population[i]
  data_type <- sample_table$data_type[i]
  if (str_detect(data_type,"pe")){
    path <- str_c("../angsd/heterozygosity/", sample_seq_id,  "_bt2_gadMor3_sorted_dedup_overlapclipped_realigned_mindp2_maxdp10_minq20_minmapq30")
    path_stringent <- str_c("../angsd/heterozygosity/", sample_seq_id,  "_bt2_gadMor3_sorted_dedup_overlapclipped_realigned_mindp2_maxdp10_minq33_minmapq30")
  } else {
    path <- str_c("../angsd/heterozygosity/", sample_seq_id,  "_bt2_gadMor3_sorted_dedup_realigned_mindp2_maxdp10_minq20_minmapq30")
    path_stringent <- str_c("../angsd/heterozygosity/", sample_seq_id,  "_bt2_gadMor3_sorted_dedup_realigned_mindp2_maxdp10_minq33_minmapq30")
  }
  het_relaxed <- read_delim(str_c(path, ".ml"), col_names = F, delim = " ") %>% 
    transmute(n_sites=(X1+X2+X3), n_snp=X2, het=n_snp/n_sites) %>%
    mutate(sample_id=sample_id, population=population, data_type=data_type, tran="Including transitions", filter="relaxed")
  het_relaxed_notrans <- read_delim(str_c(path, "_notrans.ml"), col_names = F, delim = " ") %>% 
    transmute(n_sites=(X1+X2+X3), n_snp=X2, het=n_snp/n_sites) %>%
    mutate(sample_id=sample_id, population=population, data_type=data_type, tran="Excluding transitions", filter="relaxed")
  het_stringent <- read_delim(str_c(path_stringent, ".ml"), col_names = F, delim = " ") %>% 
    transmute(n_sites=(X1+X2+X3), n_snp=X2, het=n_snp/n_sites) %>%
    mutate(sample_id=sample_id, population=population, data_type=data_type, tran="Including transitions", filter="stringent")
  het_stringent_notrans <- read_delim(str_c(path_stringent, "_notrans.ml"), col_names = F, delim = " ") %>% 
    transmute(n_sites=(X1+X2+X3), n_snp=X2, het=n_snp/n_sites) %>%
    mutate(sample_id=sample_id, population=population, data_type=data_type, tran="Excluding transitions", filter="stringent")
  het_combined <- bind_rows(het_relaxed, het_relaxed_notrans, het_stringent, het_stringent_notrans)
  if(i==1){
    het_final <- het_combined
  } else {
    het_final <- bind_rows(het_final, het_combined)
  }
}
het_per_ind <- het_final %>%
  unite(col = type, tran, filter, sep = " ") %>%
  dplyr::select(sample_id, population, data_type, type, het) %>%
  pivot_wider(names_from = type, values_from = het)
## Calculate average heterozygosity while filtering out inversions
set.seed(42)
het_final %>%
  ggplot(aes(x=population, y=het)) +
  geom_boxplot(outlier.alpha = 0) +
  geom_jitter(aes(color=data_type), height = 0, size=0.8) +
  facet_grid(filter~tran) +
  coord_flip() +
  theme_cowplot() +
  theme(panel.background=element_rect(colour="black", size=0.8))
het_final %>%
  ggplot(aes(x=n_sites, y=het, color=data_type)) +
  geom_point(height = 0, size=1) +
  geom_smooth(se = F, color="black", aes(group=data_type)) +
  facet_grid(filter~tran) +
  theme_cowplot() +
  theme(panel.background=element_rect(colour="black", size=0.8))
```

###### Effect of excluding transitions

```{r eval=TRUE, fig.width=8, fig.height=4}
set.seed(42)
delta_het <- het_final %>%
  dplyr::select(het, sample_id, population, data_type, tran, filter) %>%
  pivot_wider(names_from = tran,  values_from = het) %>%
  mutate(delta = `Excluding transitions`-`Including transitions`,
         degradation = ifelse(population %in% c("UUM2010", "NAR2008", "ATP2011") & data_type=="se", "more degraded", "less degraded"),
         type=str_c(data_type, degradation, sep = "\n"))
delta_het %>%
  ggplot(aes(x=type, y=delta)) +
  geom_boxplot(outlier.alpha = 0) +
  geom_jitter(height = 0) +
  facet_wrap(~filter) +
  theme_cowplot()
```


```{r eval=TRUE, fig.width=6, fig.height=4}
set.seed(42)
delta_het %>%
  filter(filter=="stringent") %>%
  ggplot(aes(x=type, y=delta)) +
  geom_boxplot(outlier.alpha = 0) +
  geom_jitter(height = 0) +
  #annotate(geom = "text", label="p-value=0.014", x=1.5, y=0.002) +
  scale_x_discrete(labels=c("NextSeq-150PE\n(well-preserved)\n", "HiSeq-125SE\n(well-preserved)\n", "HiSeq-125SE\n(degraded)\n")) +
  ylab("change in heterozygosity\nafter excluding transitions") +
  theme_cowplot() +
  theme(axis.title.x = element_blank())

## Only including the three populations where samples are split based on degradation level
delta_het %>%
  filter(filter=="stringent", population %in% c("UUM2010", "NAR2008", "ATP2011")) %>%
  ggplot(aes(x=type, y=delta)) +
  geom_boxplot(outlier.alpha = 0) +
  geom_jitter(height = 0) +
  #annotate(geom = "text", label="p-value=0.014", x=1.5, y=0.002) +
  scale_x_discrete(labels=c("NextSeq-150PE\n(well-preserved)\n", "HiSeq-125SE\n(degraded)\n")) +
  ylab("change in heterozygosity\nafter excluding transitions") +
  ylim(c(NA, 0)) +
  theme_cowplot() +
  theme(axis.title.x = element_blank())
```

```{r eval=TRUE, fig.width=6, fig.height=5}
set.seed(42)
p_b <- delta_het %>%
  filter(filter=="stringent") %>%
  ggstatsplot::ggbetweenstats(x = type, y = delta,  type = "p", p.adjust.method = "holm") +
  scale_x_discrete(labels=c("NextSeq-150PE\n(well-preserved)\n", "HiSeq-125SE\n(well-preserved)\n", "HiSeq-125SE\n(degraded)\n")) +
  ylab("change in heterozygosity estimates \nafter excluding transitions") +
  xlab("sample type")
print(p_b)
```

This shows that DNA damage has an stronger effect on SE samples (which are more degraded). 

###### Effect of having a stringent filter

```{r eval=TRUE, fig.width=6, fig.height=4}
set.seed(42)
delta_het_filter <- het_final %>%
  dplyr::select(het, sample_id, population, data_type, tran, filter) %>%
  pivot_wider(names_from = filter,  values_from = het) %>%
  mutate(delta = `stringent`- `relaxed`) 
delta_het_filter %>%
  ggplot(aes(x=data_type, y=delta)) +
  geom_boxplot(outlier.alpha = 0) +
  geom_jitter(height = 0) +
  geom_hline(yintercept = 0, color="red") +
  facet_wrap(~tran, scales = "free_y") +
  theme_cowplot()
t.test(filter(het_per_ind, data_type=="se")$`Including transitions relaxed`,
       filter(het_per_ind, data_type=="se")$`Including transitions stringent`,
       paired=TRUE)$p.value
t.test(filter(het_per_ind, data_type=="pe")$`Including transitions relaxed`,
       filter(het_per_ind, data_type=="pe")$`Including transitions stringent`,
       paired=TRUE)$p.value
```

```{r eval=TRUE, fig.width=5, fig.height=4}
set.seed(42)
delta_het_filter %>%
  filter(tran=="Including transitions") %>%
  ggplot(aes(x=data_type, y=delta)) +
  geom_hline(yintercept=0, color="red") +
  geom_boxplot(outlier.alpha = 0) +
  geom_jitter(height = 0) +
  annotate(geom = "text", label="p-value<0.0001", x=1, y=0.0005) +
  annotate(geom = "text", label="p-value=0.008", x=2, y=0.0005) +
  scale_x_discrete(labels=c("NextSeq-150PE\n(more biased\nbase quality)", "HiSeq-125SE\n(less biased\nbase quality)")) +
  ylab("change in heterozygosity with\na stringent mapping quality filter") +
  theme_cowplot() +
  theme(axis.title.x = element_blank())
```

```{r eval=TRUE, fig.width=12, fig.height=5}
set.seed(42)
p_a <- het_final %>%
  mutate(data_type=ifelse(data_type=="pe", "NextSeq-150PE (more miscalibration in base quality scores)", "HiSeq-125SE (less miscalibration in base quality scores)")) %>%
  mutate(filter=ifelse(filter=="relaxed", "relaxed base quality filter (minQ = 20)", "stringent base quality filter (minQ = 33)")) %>%
  filter(tran=="Including transitions") %>%
  ggstatsplot::grouped_ggwithinstats(
  x = filter,
  y = het,
  #type = "np", # non-parametric statistics
  xlab = element_blank(),
  ylab = "estimated heterozygosity",
  grouping.var = data_type,
  #ggtheme = theme_cowplot(),
  bf.message = FALSE
  )
print(p_a)
```

A more stringent filter decreases heterozygosity estimate of PE samples but increases that of the the SE samples (very slightly).

###### PE vs. SE 

```{r eval=TRUE, fig.width=10, fig.height=8}
## Within KNG2011 only
set.seed(42)
het_final %>%
  filter(str_detect(sample_id, "KNG2011")) %>%
  ggplot(aes(x=data_type, y=het)) +
  geom_boxplot(outlier.alpha = 0) +
  geom_jitter(height = 0) +
  facet_grid(filter~tran, scales = "free") +
  theme_cowplot() +
  theme(panel.background=element_rect(colour="black", size=0.8))
## PE vs SE before filtering and excluding transitions
t.test(filter(het_final, str_detect(sample_id, "KNG2011"), filter=="relaxed", tran == "Including transitions", data_type=="se")$het,
       filter(het_final, str_detect(sample_id, "KNG2011"), filter=="relaxed", tran == "Including transitions", data_type=="pe")$het)$p.value
## PE vs SE after filtering and excluding transitions
t.test(filter(het_final, str_detect(sample_id, "KNG2011"), filter=="stringent", tran == "Excluding transitions", data_type=="se")$het,
       filter(het_final, str_detect(sample_id, "KNG2011"), filter=="stringent", tran == "Excluding transitions", data_type=="pe")$het)$p.value
## With all populations, controlling for population
##lm(`Including transitions`~population + data_type, data = filter(delta_het, filter=="relaxed")) %>% summary()
##lm(`Including transitions`~population + data_type, data = filter(delta_het, filter=="stringent")) %>% summary()
##lm(`Excluding transitions`~population + data_type, data = filter(delta_het, filter=="relaxed")) %>% summary()
##lm(`Excluding transitions`~population + data_type, data = filter(delta_het, filter=="stringent")) %>% summary()
```

This shows that after excluding transitions and using stringent depth and quality filters, batch effect on heterozygosity estimate is significantly reduced.

## Base substitutions frequency in private SNPs

```{r eval=T, fig.width=6, fig.height=5}
maf_se <- read_tsv("../angsd/popminind20/se_global_snp_list_bam_list_realigned_mindp46_maxdp184_minind20_minq20_popminind20.mafs.gz") %>%
  transmute(lg = chromo, position = position, major=major, minor = minor, se_maf = knownEM, se_nind=nInd)
maf_pe <- read_tsv("../angsd/popminind20/pe_global_snp_list_bam_list_realigned_mindp46_maxdp184_minind20_minq20_popminind20.mafs.gz")%>%
  transmute(lg = chromo, position = position, major=major, minor = minor, pe_maf = knownEM, pe_nind=nInd)
maf_joined <- inner_join(maf_se, maf_pe) %>%
  mutate(delta = abs(se_maf- pe_maf))
p_c <- bind_rows((filter(maf_joined, pe_maf<0.01 | pe_maf>0.99) %>% filter(se_maf>0.1 & se_maf<0.9) %>% transmute(major=major, minor=minor, batch = "HiSeq-125SE\n(more degraded,\nless base calling bias)")),
          (filter(maf_joined, se_maf<0.01 | se_maf>0.99) %>% filter(pe_maf>0.1 & pe_maf<0.9) %>% transmute(major=major, minor=minor, batch = "NextSeq-150PE\n(well-preserved,\nmore base calling bias)"))) %>%
  mutate(base_substitution = str_c(major, "-to-", minor)) %>%
  mutate(base_substitution = case_when(
    base_substitution %in% c("A-to-C", "T-to-G") ~ "A-to-C\nT-to-G", 
    base_substitution %in% c("A-to-G", "T-to-C") ~ "A-to-G\nT-to-C", 
    base_substitution %in% c("A-to-T", "T-to-A") ~ "A-to-T\nT-to-A", 
    base_substitution %in% c("C-to-A", "G-to-T") ~ "C-to-A\nG-to-T", 
    base_substitution %in% c("C-to-G", "G-to-C") ~ "C-to-G\nG-to-C", 
    base_substitution %in% c("C-to-T", "G-to-A") ~ "C-to-T\nG-to-A"
  )) %>%
  group_by(base_substitution, batch) %>% 
  count() %>% 
  ungroup() %>% 
  group_by(batch) %>%
  mutate(frequency = n / sum(n)) %>%
  ungroup() %>%
  ggplot(aes(x=base_substitution, y=frequency, fill=batch, group=batch)) +
  #geom_line() + 
  #geom_point() +
  geom_col(position = "dodge", color="black")+
  scale_fill_viridis_d(begin = 0.25, end="0.75") +
  ylim(c(NA, 0.3)) +
  labs(x="type of base substitution", y="frequency") +
  theme_ggstatsplot() +
  theme(legend.position = "top",
        panel.grid.major.y =element_line(color="grey"))
print(p_c)
```

```{r eval=T, fig.width=6, fig.height=5}
p_c_2 <- bind_rows((filter(maf_joined, pe_maf<0.01 | pe_maf>0.99) %>% filter(se_maf>0.1 & se_maf<0.9) %>% transmute(major=major, minor=minor, batch = "HiSeq-125SE\n(more degraded,\nless base calling bias)")),
          (filter(maf_joined, se_maf<0.01 | se_maf>0.99) %>% filter(pe_maf>0.1 & pe_maf<0.9) %>% transmute(major=major, minor=minor, batch = "NextSeq-150PE\n(well-preserved,\nmore base calling bias)"))) %>%
  mutate(base_substitution = str_c(major, "-to-", minor)) %>%
  mutate(base_substitution = case_when(
    base_substitution %in% c("A-to-C", "T-to-G") ~ "A-to-C & T-to-G", 
    base_substitution %in% c("A-to-G", "T-to-C") ~ "A-to-G & T-to-C", 
    base_substitution %in% c("A-to-T", "T-to-A") ~ "A-to-T & T-to-A", 
    base_substitution %in% c("C-to-A", "G-to-T") ~ "C-to-A & G-to-T", 
    base_substitution %in% c("C-to-G", "G-to-C") ~ "C-to-G & G-to-C", 
    base_substitution %in% c("C-to-T", "G-to-A") ~ "C-to-T & G-to-A"
  )) %>%
  arrange(base_substitution) %>%
  mutate(base_substitution = factor(base_substitution)) %>%
  ggbarstats(
  x = base_substitution,
  y = batch,
  xlab = "Batch",
  legend.title = "Base substitution",
  palette = "Set2"
  )
print(p_c_2)
```

## Assemble Figure 3

```{r eval=TRUE, fig.width=12, fig.height=10}
bottom <- cowplot::plot_grid(p_b, p_c_2, nrow = 1, labels = c("B", "C"))
figure <- cowplot::plot_grid(p_a, bottom, nrow = 2, labels = c("A", NA), rel_heights = c(1, 1))
print(figure)
```

```{r eval=TRUE, fig.width=12, fig.height=10}
bottom <- cowplot::plot_grid(p_b, p_c, nrow = 1, labels = c("B", "C"))
figure <- cowplot::plot_grid(p_a, bottom, nrow = 2, labels = c("A", NA), rel_heights = c(1, 1))
print(figure)
```